### ISSUE TITLE
Questions regarding Human SL model

### ISSUE THREAD

**Issue by pnprog** [lightvector/KataGo#1074]
Thanks for releasing the "Human SL", I think there is a lot of potential to use that model, and I have been experimenting a lot with it recently.

I have some question regarding specifically the model file used (v1.15.0/b18c384nbt-humanv0.bin.gz):

- How is this file built compared to classic models? is it like a collection of models (one for each profile) stacked together? Is there a tool I can use to "inspect" it?
- What is the "block size" of this model? like 40-block? 60-block?
- How easy or hard would it be to expand the number of profiles that make that model? For instance, if I wanted to add the profiles `preaz_30k` to `preaz_21k` (or `proyear_1700` to `proyear_1799`) to the existing list of models? (given that I already have an existing trained model, or the game collections to train it if necessary)

One last question, regarding the profiles made from older games:

- For profiles that have been trained on games record from before the current komi value, or even before the introduction of komi (for instance `proyear_1900`), can we expect the model to be at a disadvantage because of that? For instance, could the model believe it is loosing the game as white, because it "does not know" it has a 7.5pt komi?

Thanks a lot!

**Comment by HackYardo** [comment]
1. the biggest difference between classic models and humansl models is the **training data**. hsl models use human Go game records to train, so its **style** is more human-like and its **power** is less than a superhuman. if you want to "inspect" it, watch the file `b18c384nbt-humanv0.ckpt` from KataGo 1.15.0 release webpage
2. block is the size of a neural network model, that is to say **the capacity of a container** (nnm), *see also [LeelachessZero](https://lczero.org/dev/backend/nn/)*
3. you should understand the meaning of these parameters, one can not reach 30k unless he or she intend to loss every Go games
4. komi is for final scores, not for power, e.g. Lee Se-dol can not reach 30k even if the komi is 180, because he plays Black as many as plays White.

**Comment by pnprog** [comment]
Hello,

Thanks for your answers!

> 1. the biggest difference between classic models and humansl models is the **training data**. hsl models use human Go game records to train, so its **style** is more human-like and its **power** is less than a superhuman. if you want to "inspect" it, watch the file `b18c384nbt-humanv0.ckpt` from KataGo 1.15.0 release webpage

Yes, i am clear about that. But what I try to understand is something different:

For comparison, the current best network is `kata1-b28c512nbt-s9225415424-d4866680860.bin.gz`, and plays at super human strength, and that's it.

But `b18c384nbt-humanv0.bin.gz` is different, since it can plays both at human 9d strength, or human 20k strength, or like a pro from 1800. Considering the variants `preaz_*` and `rank_*` that ranges from `20k` to `9d`, the variant `proyear_*` that ranges from `1800` to `2023`, it's like this network includes `29+29+223=281` different `sub-networks` (I don't don't know of to call that) that all play with their own strength and style. And there are also `preaz_{BR}_{WR}` and `rank_{BR}_{WR}`...

I would have expected each of those variants to correspond to a specific network, trained on dedicated training data. And so, I would have expected `b18c384nbt-humanv0.bin.gz` to be like a **collection of all those networks packed together**. But the file is only about 96.7Mo and is about the same weight as any other 18 blocks networks (for comparison, `kata1-b18c384nbt-s9996604416-d4316597426`, the strongest 18 blocks network is 93.4Mo).

So what am I missing here?

>     2. block is the size of a neural network model, that is to say **the capacity of a container** (nnm), _see also [LeelachessZero](https://lczero.org/dev/backend/nn/)_

Yes, sorry for my poor phrasing, I was asking for the actual block size of `b18c384nbt-humanv0.bin.gz` here. So I checked the `ckpt` file with pytorch, and learnt it is a 18 block network. And then I realized the filename starting with `b18` was already making that clear...

>     3. you should understand the meaning of these parameters, one can not reach 30k unless he or she intend to loss every Go games

That's just me being not clear here. I try to understand how the special `b18c384nbt-humanv0.bin.gz` function. To give a more realistic use case: **How easy or hard would it be to expand the number of profiles that make that model? For instance, if I wanted to add the profile `proyear_2024`?**

Why would I ask that: I realized one advantage of the `b18c384nbt-humanv0.bin.gz` is that **it seems** I can run one instance of the katago analysis engine (for instance behind a server), and then switch the network "on demand" by simply adjusting `humanSLProfile`, and without having to restart the analysis engine (and keeping the memory footprint very low). Of course this is limited to the existing profiles (`preaz_*`, `proyear_*`...). But could this network be extended to add a few super human strength 18 blocks files as well for instance? (But clearly there is something I am missing here)


>     4. komi is for final scores, not for power, e.g. Lee Se-dol can not reach 30k even if the komi is 180, because he plays Black as many as plays White.

To give a concrete example, let's consider the `humanWinrate` field of the `rootInfo` field. Can I assume that if using `proyear_1800` as `humanSLProfile`, this value will correspond to a winrate that consider no komi for white? (since in 1800 there was no komi). And same remark for `humanScoreMean` or `humanScoreStdev`.

Now, if I run katago analysis this way:

> An alternative way: pass `-model b18c384nbt-humanv0.bin.gz` instead of KataGo's normal model, using the human model exclusively.

In that case, would `rawWinrate` and `rawScoreSelfplay` in `rootInfo` have similar bias as well?

Thanks for your patience :)

**Comment by lightvector** [comment]
> I would have expected each of those variants to correspond to a specific network, trained on dedicated training data. And so, I would have expected b18c384nbt-humanv0.bin.gz to be like a collection of all those networks packed together. But the file is only about 96.7Mo and is about the same weight as any other 18 blocks networks (for comparison, kata1-b18c384nbt-s9996604416-d4316597426, the strongest 18 blocks network is 93.4Mo).
> So what am I missing here?

It's exactly the same as why the same network can play different komis, or why the same network can play different board sizes, or why the same network can play the opening and the midgame and the endgame, or why the same network can play josekis that arise from the 4-4 point as arise from the 3-4 point. In theory we could have separate networks for different komis, or for different board sizes, or for phases of the game, or even for different josekis. But it's possible to train one network that handles all of them at once, as long as the training data contains all of them.

> That's just me being not clear here. I try to understand how the special b18c384nbt-humanv0.bin.gz function. To give a more realistic use case: How easy or hard would it be to expand the number of profiles that make that model? For instance, if I wanted to add the profile proyear_2024?

You would need to obtain all necessary training data for 2024 (e.g. all the 2024 pro games) and learn how to use all the KataGo training scripts and python code. The training code tends to be oriented towards programmers/engineers, so if you're not the kind of person who is already familiar with cloning github repos and hacking on code and scripts, or working with datasets and basic machine learning setups, you might have to learn. You would also need a decently strong GPU. Additionally, with sufficiently much training on 2024, the net would eventually likely forget how to make predictions on all ranks and settings other than "pro, 2024" due to the phenomenon of "catastrophic forgetting" (google if you don't know about this) in neural nets, unless you also continued to train it on all the other things you didn't want it to forget.

> To give a concrete example, let's consider the humanWinrate field of the rootInfo field. Can I assume that if using proyear_1800 as humanSLProfile, this value will correspond to a winrate that consider no komi for white? (since in 1800 there was no komi). And same remark for humanScoreMean or humanScoreStdev.

KataGo's nets are always told exactly what the komi is, both when they are trained, and when they are playing. This is why they can make different predictions when you change the komi. When the net was trained on games from 1800, if there was no komi, then then it was also told that there was no komi (unless there were errors in how those old games were recorded in the GoGoD dataset, which is the dataset that was used for old games). The komi is never something that is assumed by the net - it is always what you say it is in the GUI or SGF file, and then that value is told to the neural net. If you tell the neural net that the komi is some komi value for which there exists no data (e.g. komi of 7.5 in an era where there were no games with such a komi), it will make an arbitrary extrapolated guess of what the outputs would be with such a value that will often be reasonable, but are not guaranteed to be, because that's what neural nets always do when asked for something outside of the range of their training.
